ruma-federation-api/src/transactions/send_transaction_message/edu.rs

```diff
@@ -0,0 +1,288 @@
+    #[serde(rename = "m.presence")]
+    Presence(PresenceContent),
+
+    /// An EDU representing receipt updates for users of the sending homeserver.
+    #[serde(rename = "m.receipt")]
+    Receipt(ReceiptContent),
+
+    /// A typing notification EDU for a user in a room.
+    #[serde(rename = "m.typing")]
+    Typing(TypingContent),
+
+    /// An EDU that lets servers push details to each other when one of their users adds
+    // a new device to their account, required for E2E encryption to correctly target the
+    // current set of devices for a given user.
+    #[serde(rename = "m.device_list_update")]
+    DeviceListUpdate(DeviceListUpdateContent),
+
+    /// An EDU that lets servers push send events directly to a specific device on a
+    // remote server - for instance, to maintain an Olm E2E encrypted message channel
+    // between a local and remote device.
+    #[serde(rename = "m.direct_to_device")]
+    DirectToDevice(DirectDeviceContent),
+
+    #[doc(hidden)]
+    Custom(JsonValue),
+}
+
+#[derive(Debug, Deserialize)]
+struct EduTpeDeHelper {
+    /// The message type field
+    edu_type: String,
+    content: Box<RawJsonValue>,
+}
+
+impl<'de> de::Deserialize<'de> for Edu {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let EduTpeDeHelper { edu_type, ref content } = from_raw_json_value(&json)?;
+
+        Ok(match edu_type.as_ref() {
+            "m.presence" => Self::Presence(from_raw_json_value(content)?),
+            "m.receipt" => Self::Receipt(from_raw_json_value(content)?),
+            "m.typing" => Self::Typing(from_raw_json_value(content)?),
+            "m.device_list_update" => Self::DeviceListUpdate(from_raw_json_value(content)?),
+            "m.direct_to_device" => Self::DirectToDevice(from_raw_json_value(content)?),
+            _ => Self::Custom(from_raw_json_value(content)?),
+        })
+    }
+}
+
+/// The content for "m.presence" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct PresenceContent {
+    /// A list of presence updates that the receiving server is likely to be interested in.
+    pub push: Vec<PresenceUpdate>,
+}
+
+/// An update to the presence of a user.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct PresenceUpdate {
+    /// The user ID this presence EDU is for.
+    pub user_id: UserId,
+
+    /// The presence of the user.
+    pub presence: PresenceState,
+
+    /// An optional description to accompany the presence.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub status_msg: Option<String>,
+
+    /// The number of milliseconds that have elapsed since the user last did something.
+    pub last_active_ago: UInt,
+
+    /// Whether or not the user is currently active.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub currently_active: Option<bool>,
```

Comment:
When do we add `#[serde(default, ...` to a field. If the spec says they are optional don't we always need to default initialize them in case we get JSON without that field.

Reply:
If there is no semantic difference between `None` and one of the `Some(_)` cases we shouldn't use `Option` and add `serde(default [= "fn"])`. `Option` basically has an implicit `serde(default)`, so this code would work without the field being present in the JSON in when deserializing.

Reply:
I'm pretty sure I got this all figured out, maybe worth a check.